File: src/lockfile.c
#include "lockfile.h"
#include "util.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define LOCKFILE_NAME "rosie.lock"
#define LOCKFILE_LINE_MAX 2048
#define LOCKFILE_VERSION 1

static void lock_entry_free(LockEntry *e) {
    if (!e) return;
    spm_free(e->skill_name);
    spm_free(e->source);
    spm_free(e->ref);
    spm_free(e->sha);
    spm_free(e->installed_at);
    e->skill_name = NULL;
    e->source = NULL;
    e->ref = NULL;
    e->sha = NULL;
    e->installed_at = NULL;
}

void lockfile_free(Lockfile *lf) {
    if (!lf) return;
    for (int i = 0; i < lf->count; i++) {
        lock_entry_free(&lf->entries[i]);
    }
    spm_free(lf->entries);
    spm_free(lf->path);
    spm_free(lf);
}

static void lockfile_grow(Lockfile *lf) {
    if (lf->count < lf->capacity) return;
    int new_cap = lf->capacity == 0 ? 8 : lf->capacity * 2;
    lf->entries = spm_realloc(lf->entries, new_cap * sizeof(LockEntry));
    lf->capacity = new_cap;
}

Lockfile *lockfile_load(const char *dir) {
    Lockfile *lf = spm_malloc(sizeof(Lockfile));
    lf->entries = NULL;
    lf->count = 0;
    lf->capacity = 0;
    lf->path = path_join(dir, LOCKFILE_NAME);

    FILE *fp = fopen(lf->path, "r");
    if (!fp) {
        // Missing file is fine — return empty lockfile
        return lf;
    }

    char line[LOCKFILE_LINE_MAX];
    while (fgets(line, sizeof(line), fp)) {
        char *trimmed = str_trim(line);
        if (trimmed[0] == '\0' || trimmed[0] == '#') continue;

        char skill_name[256], source[1024], ref[256], sha[64], ts[64], pin[16], kind[16];
        pin[0] = '\0';
        kind[0] = '\0';
        int n = sscanf(trimmed, "%255s %1023s %255s %63s %63s %15s %15s",
                       skill_name, source, ref, sha, ts, pin, kind);
        if (n < 5) {
            log_debug("Skipping malformed lockfile line: %s", trimmed);
            continue;
        }

        lockfile_grow(lf);
        LockEntry *e = &lf->entries[lf->count++];
        e->skill_name = str_dup(skill_name);
        e->source = str_dup(source);
        e->ref = str_dup(ref);
        e->sha = str_dup(sha);
        e->installed_at = str_dup(ts);
        // Pinned flag is optional for backwards compat; default to false (auto).
        e->pinned = (n >= 6) && (strcmp(pin, "pin") == 0);
        // Kind is optional (legacy v0 lockfiles have no 7th field). Default to skill.
        e->kind = (n >= 7 && strcmp(kind, "ref") == 0) ? LOCK_REF : LOCK_SKILL;
    }

    fclose(fp);
    return lf;
}

static int lock_entry_cmp(const void *a, const void *b) {
    const LockEntry *ea = (const LockEntry *)a;
    const LockEntry *eb = (const LockEntry *)b;
    return strcmp(ea->skill_name, eb->skill_name);
}

int lockfile_save(const Lockfile *lf) {
    if (!lf || !lf->path) return -1;

    // Sort for stable diffs. The cast away from const is safe — we only reorder
    // entries, not their string contents, and the array is owned by lf.
    if (lf->count > 1) {
        qsort(lf->entries, lf->count, sizeof(LockEntry), lock_entry_cmp);
    }

    size_t tmp_len = strlen(lf->path) + 5;
    char *tmp_path = spm_malloc(tmp_len);
    snprintf(tmp_path, tmp_len, "%s.tmp", lf->path);

    FILE *fp = fopen(tmp_path, "w");
    if (!fp) {
        log_error("Cannot create lockfile: %s", tmp_path);
        spm_free(tmp_path);
        return -1;
    }

    fprintf(fp, "# rosie-lock v%d\n", LOCKFILE_VERSION);

    for (int i = 0; i < lf->count; i++) {
        const LockEntry *e = &lf->entries[i];
        fprintf(fp, "%s %s %s %s %s %s %s\n",
                e->skill_name, e->source, e->ref, e->sha, e->installed_at,
                e->pinned ? "pin" : "auto",
                e->kind == LOCK_REF ? "ref" : "skill");
    }

    if (fclose(fp) != 0) {
        log_error("Failed writing lockfile: %s", tmp_path);
        unlink(tmp_path);
        spm_free(tmp_path);
        return -1;
    }

    if (rename(tmp_path, lf->path) != 0) {
        log_error("Cannot finalize lockfile: %s", lf->path);
        unlink(tmp_path);
        spm_free(tmp_path);
        return -1;
    }

    spm_free(tmp_path);
    return 0;
}

LockEntry *lockfile_find(const Lockfile *lf, const char *skill_name) {
    if (!lf || !skill_name) return NULL;
    for (int i = 0; i < lf->count; i++) {
        if (strcmp(lf->entries[i].skill_name, skill_name) == 0) {
            return &lf->entries[i];
        }
    }
    return NULL;
}

void lockfile_upsert(Lockfile *lf, const char *skill_name, const char *source,
                     const char *ref, const char *sha, const char *installed_at,
                     bool pinned, LockKind kind) {
    if (!lf || !skill_name) return;

    LockEntry *e = lockfile_find(lf, skill_name);
    if (e) {
        spm_free(e->source);
        spm_free(e->ref);
        spm_free(e->sha);
        spm_free(e->installed_at);
        e->source = str_dup(source);
        e->ref = str_dup(ref);
        e->sha = str_dup(sha);
        e->installed_at = str_dup(installed_at);
        e->pinned = pinned;
        e->kind = kind;
        return;
    }

    lockfile_grow(lf);
    e = &lf->entries[lf->count++];
    e->skill_name = str_dup(skill_name);
    e->source = str_dup(source);
    e->ref = str_dup(ref);
    e->sha = str_dup(sha);
    e->installed_at = str_dup(installed_at);
    e->pinned = pinned;
    e->kind = kind;
}

int lockfile_remove_entry(Lockfile *lf, const char *skill_name) {
    if (!lf || !skill_name) return 0;
    for (int i = 0; i < lf->count; i++) {
        if (strcmp(lf->entries[i].skill_name, skill_name) == 0) {
            lock_entry_free(&lf->entries[i]);
            for (int j = i; j < lf->count - 1; j++) {
                lf->entries[j] = lf->entries[j + 1];
            }
            lf->count--;
            return 1;
        }
    }
    return 0;
}

char *lockfile_now_iso8601(void) {
    time_t now = time(NULL);
    struct tm tm_utc;
    gmtime_r(&now, &tm_utc);
    char *buf = spm_malloc(32);
    strftime(buf, 32, "%Y-%m-%dT%H:%M:%SZ", &tm_utc);
    return buf;
}

File: src/lockfile.rs
// .agents/rosie.lock — whitespace-separated entries with a v1 header.
//
// The format is locked down by the regression suite; even whitespace
// changes break tests. Layout:
//
//   # rosie-lock v1
//   <name> <source> <ref> <sha> <iso8601_ts> <pin|auto> <skill|ref>
//
// Entries are sorted by name on save for stable diffs (the suite ordering
// is alphabetical).

use crate::os;
use crate::util;
use std::path::{Path, PathBuf};

pub const LOCKFILE_NAME: &str = "rosie.lock";
pub const LOCKFILE_VERSION: u32 = 1;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LockKind {
    Skill,
    Ref,
}

impl LockKind {
    fn as_str(self) -> &'static str {
        match self {
            LockKind::Skill => "skill",
            LockKind::Ref => "ref",
        }
    }
}

#[derive(Debug, Clone)]
pub struct LockEntry {
    pub skill_name: String,
    pub source: String,
    pub ref_: String,
    pub sha: String,
    pub installed_at: String,
    pub pinned: bool,
    pub kind: LockKind,
}

#[derive(Debug, Clone)]
pub struct Lockfile {
    pub entries: Vec<LockEntry>,
    pub path: PathBuf,
}

impl Lockfile {
    /// Load `<dir>/rosie.lock`. Returns an empty lockfile if the file is
    /// missing — same as the C version.
    pub fn load(dir: &Path) -> Self {
        let path = dir.join(LOCKFILE_NAME);
        let mut entries = Vec::new();

        if let Ok(contents) = os::read_to_string(&path) {
            for line in contents.lines() {
                let trimmed = line.trim();
                if trimmed.is_empty() || trimmed.starts_with('#') {
                    continue;
                }
                let parts: Vec<&str> = trimmed.split_whitespace().collect();
                if parts.len() < 5 {
                    crate::log::debug(&format!("skipping malformed lockfile line: {trimmed}"));
                    continue;
                }
                let pinned = parts.get(5).copied() == Some("pin");
                let kind = if parts.get(6).copied() == Some("ref") {
                    LockKind::Ref
                } else {
                    LockKind::Skill
                };
                entries.push(LockEntry {
                    skill_name: parts[0].to_string(),
                    source: parts[1].to_string(),
                    ref_: parts[2].to_string(),
                    sha: parts[3].to_string(),
                    installed_at: parts[4].to_string(),
                    pinned,
                    kind,
                });
            }
        }

        Lockfile { entries, path }
    }

    pub fn find(&self, skill_name: &str) -> Option<&LockEntry> {
        self.entries.iter().find(|e| e.skill_name == skill_name)
    }

    pub fn find_mut(&mut self, skill_name: &str) -> Option<&mut LockEntry> {
        self.entries.iter_mut().find(|e| e.skill_name == skill_name)
    }

    /// Insert or replace by skill name.
    #[allow(clippy::too_many_arguments)]
    pub fn upsert(
        &mut self,
        skill_name: &str,
        source: &str,
        ref_: &str,
        sha: &str,
        installed_at: &str,
        pinned: bool,
        kind: LockKind,
    ) {
        if let Some(e) = self.find_mut(skill_name) {
            e.source = source.to_string();
            e.ref_ = ref_.to_string();
            e.sha = sha.to_string();
            e.installed_at = installed_at.to_string();
            e.pinned = pinned;
            e.kind = kind;
            return;
        }
        self.entries.push(LockEntry {
            skill_name: skill_name.to_string(),
            source: source.to_string(),
            ref_: ref_.to_string(),
            sha: sha.to_string(),
            installed_at: installed_at.to_string(),
            pinned,
            kind,
        });
    }

    /// Remove an entry by name. Returns true if present.
    pub fn remove(&mut self, skill_name: &str) -> bool {
        let before = self.entries.len();
        self.entries.retain(|e| e.skill_name != skill_name);
        self.entries.len() != before
    }

    /// Atomic save: write to .tmp then rename. Entries are sorted by name
    /// first for stable diffs.
    pub fn save(&mut self) -> Result<(), os::OsError> {
        self.entries.sort_by(|a, b| a.skill_name.cmp(&b.skill_name));

        let mut out = String::new();
        out.push_str(&format!("# rosie-lock v{}\n", LOCKFILE_VERSION));
        for e in &self.entries {
            out.push_str(&format!(
                "{} {} {} {} {} {} {}\n",
                e.skill_name,
                e.source,
                e.ref_,
                e.sha,
                e.installed_at,
                if e.pinned { "pin" } else { "auto" },
                e.kind.as_str(),
            ));
        }

        let tmp = util::tmp_path_for(&self.path);
        os::write(&tmp, out.as_bytes())?;
        os::rename(&tmp, &self.path)
    }
}

/// Current UTC time as ISO 8601 ("2026-05-02T14:32:18Z").
pub fn now_iso8601() -> String {
    let unix = os::now_unix_seconds();
    iso8601_from_unix(unix)
}

/// Pure-function ISO 8601 formatter (broken out for testability).
pub fn iso8601_from_unix(unix_seconds: i64) -> String {
    let dt = time::OffsetDateTime::from_unix_timestamp(unix_seconds)
        .unwrap_or_else(|_| time::OffsetDateTime::UNIX_EPOCH);
    let (y, m, d) = (dt.year(), dt.month() as u8, dt.day());
    let (hh, mm, ss) = (dt.hour(), dt.minute(), dt.second());
    format!("{y:04}-{m:02}-{d:02}T{hh:02}:{mm:02}:{ss:02}Z")
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;

    #[test]
    fn round_trip() {
        let dir = std::env::temp_dir().join(format!("rosie-lockfile-test-{}", std::process::id()));
        std::fs::create_dir_all(&dir).unwrap();
        let path = dir.join(LOCKFILE_NAME);
        let mut f = std::fs::File::create(&path).unwrap();
        write!(
            f,
            "# rosie-lock v1\n\
             alpha fake/org main - 2025-01-01T00:00:00Z auto skill\n\
             beta fake/org v1.0.0 abcdef 2025-01-02T00:00:00Z pin ref\n"
        )
        .unwrap();
        drop(f);

        let mut lf = Lockfile::load(&dir);
        assert_eq!(lf.entries.len(), 2);
        assert_eq!(lf.entries[0].skill_name, "alpha");
        assert!(!lf.entries[0].pinned);
        assert_eq!(lf.entries[0].kind, LockKind::Skill);
        assert_eq!(lf.entries[1].kind, LockKind::Ref);
        assert!(lf.entries[1].pinned);

        // Save and re-load should produce identical content.
        lf.save().unwrap();
        let contents = std::fs::read_to_string(&path).unwrap();
        assert!(contents.starts_with("# rosie-lock v1\n"));
        assert!(contents.contains("alpha fake/org main - 2025-01-01T00:00:00Z auto skill\n"));
        assert!(contents.contains("beta fake/org v1.0.0 abcdef 2025-01-02T00:00:00Z pin ref\n"));

        std::fs::remove_dir_all(&dir).ok();
    }

    #[test]
    fn iso8601_known_value() {
        // 2025-01-01T00:00:00Z
        assert_eq!(iso8601_from_unix(1735689600), "2025-01-01T00:00:00Z");
    }
}
